python/tests/test_combinatorics.py

```diff
@@ -0,0 +1,472 @@
+        by the canonical labelling we assign t the minimum label.
+        Then, for every combination of k - 1 elements from the remaining
+        x * k - 1 labels given to G, we give t that combination.
+        Then, we cross all labellings of t with all the labellings
+        of the rest of the trees in G labelled with the remaining
+        (x - 1) * k labels.
+    """
+    if labels is None:
+        labels = list(range(tree.num_leaves))
+
+    if tree.is_leaf():
+        assert len(labels) == 1
+        yield RankTree(label=labels[0])
+    else:
+        groups = tree.group_children_by_shape()
+        for labeled_children in label_all_groups(groups, labels):
+            yield RankTree(children=labeled_children)
+
+
+def label_all_groups(groups, labels):
+    if len(groups) == 0:
+        yield []
+    else:
+        g, rest = groups[0], groups[1:]
+        x = len(g)
+        k = g[0].num_leaves
+        for g_labels in itertools.combinations(labels, x * k):
+            rest_labels = comb.set_minus(labels, g_labels)
+            for labeled_g in label_tree_group(g, g_labels):
+                for labeled_rest in label_all_groups(rest, rest_labels):
+                    yield labeled_g + labeled_rest
+
+
+def label_tree_group(trees, labels):
+    if len(trees) == 0:
+        assert len(labels) == 0
+        yield []
+    else:
+        first, rest = trees[0], trees[1:]
+        k = first.num_leaves
+        min_label = labels[0]
+        for first_other_labels in itertools.combinations(labels[1:], k - 1):
+            first_labels = [min_label] + list(first_other_labels)
+            rest_labels = comb.set_minus(labels, first_labels)
+            for labeled_first in all_labellings(first, first_labels):
+                for labeled_rest in label_tree_group(rest, rest_labels):
+                    yield [labeled_first] + labeled_rest
+
+
+class TestCombination(unittest.TestCase):
+    def test_combination_with_replacement_rank_unrank(self):
+        for n in range(9):
+            for k in range(n):
+                nums = list(range(n))
+                combs = itertools.combinations_with_replacement(nums, k)
+                for exp_rank, c in enumerate(combs):
+                    c = list(c)
+                    actual_rank = comb.Combination.with_replacement_rank(c, n)
+                    self.assertEqual(actual_rank, exp_rank)
+                    unranked = comb.Combination.with_replacement_unrank(exp_rank, n, k)
+                    self.assertEqual(unranked, c)
+
+    def test_combination_rank_unrank(self):
+        for n in range(11):
+            for k in range(n):
+                nums = list(range(n))
+                for rank, c in enumerate(itertools.combinations(nums, k)):
+                    c = list(c)
+                    self.assertEqual(comb.Combination.rank(c, nums), rank)
+                    self.assertEqual(comb.Combination.unrank(rank, nums, k), c)
+
+
+class TestPartition(unittest.TestCase):
```

Comment:
Would be nice leave ``rule_asc`` in it's original form, and put in a few tests against known values for it (up to 6, say). Then test separately that the ``partitions`` function is equal to ``list(rule_asc(n))[:-1]``.